### PR TITLE
sum_over_dimension and permute

### PR DESCRIPTION
**Pull request by gtdeveloper5**
add a sum_over_dimension function.
remove all old sum over functions in gtplus.
speedup permute function to use memcpy and/or multi-threading when possible.
remove all old permute functions in gtplus.

### CODE REVIEW

**Comment by inati** [comment]
I would have thought that that the permute function would be entirely memory bound so that OpenMP would not be useful.  Do we have any evidence that openMP helps here?

**Comment by gtdeveloper5** [comment]
Yes, for the 3D head testing dataset, the 3D grappa is currently done with RO decoupling. In the final permute step, without openmp part, the permute took ~150ms. With the openmp part, it took 89ms.

Since computers usually have 8-25mb L3 cache, it is a chance that different core can permute different pieces of array within this big cache.

**Comment by gtdeveloper5** [comment]
As discussed, the openmp lines are removed from permute function.

File: toolboxes/gtplus/algorithm/gtPlusOperator.h [hunk 1]
```diff
@@ -255,4 +257,70 @@ proximity(hoNDArray<T>& /*x*/, value_type /*thres*/)
     return true;
 }
 
+template<typename T>
+bool gtPlusOperator<T>::
+sumOverSecondLastDimension(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusOperator.h]: Again, don't think this function should be here.

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DT.h [hunk 2]
```diff
@@ -2786,4 +2796,204 @@ estimateJobSize(gtPlusReconWorkOrder<T>* workOrder3DT, size_t maxNumOfBytesPerJo
     return true;
 }
 
+template<typename T>
+bool gtPlusReconWorker3DT<T>::
+permuteROTo3rdDimensionFor3DRecon(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DT.h]: This function should not be here if we have a general permute

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DT.h [hunk 3]
```diff
@@ -2786,4 +2796,204 @@ estimateJobSize(gtPlusReconWorkOrder<T>* workOrder3DT, size_t maxNumOfBytesPerJo
...
+    {
+        GERROR_STREAM("Errors in gtPlusReconWorker3DT<T>::permuteROTo3rdDimensionFor3DRecon(const hoNDArray<T>& x, hoNDArray<T>& r) ... ");
+        return false;
+    }
+    return true;
+}
+
+template<typename T>
+bool gtPlusReconWorker3DT<T>::
+permuteROTo4thDimensionFor3DRecon(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DT.h]: Should not be here

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DT.h [hunk 4]
```diff
@@ -2786,4 +2796,204 @@ estimateJobSize(gtPlusReconWorkOrder<T>* workOrder3DT, size_t maxNumOfBytesPerJo
...
+    {
+        GERROR_STREAM("Errors in gtPlusReconWorker3DT<T>::permuteROTo4thDimensionFor3DRecon(const hoNDArray<T>& x, hoNDArray<T>& r) ... ");
+        return false;
+    }
+    return true;
+}
+
+template<typename T>
+bool gtPlusReconWorker3DT<T>::
+permuteROTo1stDimensionFor3DRecon(const hoNDArray<T>& x, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DT.h]: Should not be here

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTGRAPPA.h [hunk 5]
```diff
@@ -427,7 +427,7 @@ performUnwrapping(gtPlusReconWorkOrder3DT<T>* workOrder3DT, const hoNDArray<T>&
                     if ( performTiming_ ) { gt_timer3_.stop(); }
 
                     if ( performTiming_ ) { gt_timer3_.start("permuteROTo3rdDimensionFor3DRecon for aliased images ... "); }
-                    GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo3rdDimensionFor3DRecon(aliasedIm, aliasedImPermuted));
+                    GADGET_CHECK_RETURN_FALSE(this->permuteROTo3rdDimensionFor3DRecon(aliasedIm, aliasedImPermuted));
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTGRAPPA.h]: This strange call should also be replaced with a generic permute call.

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTL1SPIRITNCG.h [hunk 6]
```diff
@@ -203,21 +203,30 @@ performUnwarppingImpl(gtPlusReconWorkOrder<T>* workOrder3DT, hoNDArray<T>& kspac
         hoNDArrayMemoryManaged<T> kspaceIfftROPermuted(E1, E2, srcCHA, RO, gtPlus_mem_manager_);
 
         if ( performTiming_ ) { gt_timer3_.start("permtue RO to 4th dimension ... "); }
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo4thDimensionFor3DRecon(kspaceIfftRO, kspaceIfftROPermuted));
+        GADGET_CHECK_RETURN_FALSE(this->permuteROTo4thDimensionFor3DRecon(kspaceIfftRO, kspaceIfftROPermuted));
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTL1SPIRITNCG.h]: Again, should not be here.

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h [hunk 7]
```diff
@@ -383,7 +383,7 @@ performUnwrapping(gtPlusReconWorkOrder3DT<T>* workOrder3DT, const hoNDArray<T>&
             // hoNDArrayMemoryManaged<T> kspaceIfftROPermuted(E1, E2, srcCHA, RO, N, gtPlus_mem_manager_);
             hoNDArray<T> kspaceIfftROPermuted(E1, E2, srcCHA, RO, N);
             if ( performTiming_ ) { gt_timer3_.start("permute kspace RO to 4th dimension ... "); }
-            GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo4thDimensionFor3DRecon(kspaceIfftRO, kspaceIfftROPermuted));
+            GADGET_CHECK_RETURN_FALSE(this->permuteROTo4thDimensionFor3DRecon(kspaceIfftRO, kspaceIfftROPermuted));
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h]: Again, should be replaced with generic calls.

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h [hunk 8]
```diff
@@ -442,7 +450,7 @@ performUnwrapping(gtPlusReconWorkOrder3DT<T>* workOrder3DT, const hoNDArray<T>&
             if ( workOrder3DT->coilMap_->get_number_of_elements() > 0 )
             {
                 if ( performTiming_ ) { gt_timer3_.start("permute coil map RO to 4th dimension ... "); }
-                GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo4thDimensionFor3DRecon(*workOrder3DT->coilMap_, *coilMapPermuted));
+                GADGET_CHECK_RETURN_FALSE(this->permuteROTo4thDimensionFor3DRecon(*workOrder3DT->coilMap_, *coilMapPermuted));
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h]: I'm going to stop writing the same thing over and over, but this should be a generic call.

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h [hunk 9]
```diff
@@ -641,7 +649,7 @@ performUnwrapping(gtPlusReconWorkOrder3DT<T>* workOrder3DT, const hoNDArray<T>&
 
             // permute the unwrapped kspace
             if ( performTiming_ ) { gt_timer3_.start("permtue RO to 1st dimension ... "); }
-            GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo1stDimensionFor3DRecon(workOrder3DTJobSplit.fullkspace_, kspaceIfftRO));
+            GADGET_CHECK_RETURN_FALSE(this->permuteROTo1stDimensionFor3DRecon(workOrder3DTJobSplit.fullkspace_, kspaceIfftRO));
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h]: again

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h [hunk 10]
```diff
@@ -887,7 +895,7 @@ performUnwarppingImpl(gtPlusReconWorkOrder<T>* workOrder3DT, hoNDArray<T>& kspac
         hoNDArrayMemoryManaged<T> kspaceIfftROPermuted(E1, E2, srcCHA, RO, gtPlus_mem_manager_);
 
         if ( performTiming_ ) { gt_timer3_.start("permtue RO to 4th dimension ... "); }
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo4thDimensionFor3DRecon(kspaceIfftRO, kspaceIfftROPermuted));
+        GADGET_CHECK_RETURN_FALSE(this->permuteROTo4thDimensionFor3DRecon(kspaceIfftRO, kspaceIfftROPermuted));
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h]: again

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h [hunk 11]
```diff
@@ -981,7 +989,7 @@ performUnwarppingImpl(gtPlusReconWorkOrder<T>* workOrder3DT, hoNDArray<T>& kspac
 
         // permute the unwrapped kspace
         if ( performTiming_ ) { gt_timer3_.start("permtue RO to 1st dimension ... "); }
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo1stDimensionFor3DRecon(resDecoupled, kspaceIfftRO));
+        GADGET_CHECK_RETURN_FALSE(this->permuteROTo1stDimensionFor3DRecon(resDecoupled, kspaceIfftRO));
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconWorker3DTSPIRIT.h]: again

**Comment by gtdeveloper5** [comment]
all remaining permuteRO calls are replaced with general permute function.

File: toolboxes/gtplus/algorithm/gtPlusGRAPPA.h [hunk 12]
```diff
@@ -933,7 +935,33 @@ imageDomainKernelRO3D(const ho7DArray<T>& ker, size_t kRO, const std::vector<int
         if ( performTiming_ ) { gt_timer3_.stop(); }
 
         if ( performTiming_ ) { gt_timer3_.start("grappa 3D calibration - permute kernel dimensions to be [kE1 kE2 RO ...]  ... "); }
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo3rdDimensionFor3DRecon(kImROTemp, kImRO));
+
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusGRAPPA.h]: Looks like you have just inlined the permute here. Why don't you just call the permute function?

File: toolboxes/gtplus/algorithm/gtPlusOperator.h [hunk 13]
```diff
@@ -106,6 +106,8 @@ class gtPlusOperator
     hoNDArrayMemoryManaged<T> complexIm_Managed_;
     hoNDArrayMemoryManaged<T> res_after_apply_kernel_Managed_;
     hoNDArrayMemoryManaged<T> res_after_apply_kernel_sum_over_Managed_;
+
+    bool sumOverSecondLastDimension(const hoNDArray<T>& x, hoNDArray<T>& r);
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusOperator.h]: Why is that still here?

File: toolboxes/gtplus/algorithm/gtPlusSPIRIT.h [hunk 14]
```diff
@@ -1060,7 +1073,33 @@ imageDomainKernelRO3D(const hoNDArray<T>& ker, size_t kRO, size_t kE1, size_t kE
         if ( performTiming_ ) { gt_timer3_.stop(); }
 
         if ( performTiming_ ) { gt_timer3_.start("spirit 3D calibration - permute kernel dimensions to be [kE1 kE2 RO ...]  ... "); }
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::permuteROTo3rdDimensionFor3DRecon(kImROTemp, kImRO));
+
+        size_t N3D = ro*kConvE1*kConvE2;
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRIT.h]: Again, now it is just inlined. Don't understand what is going on here.

File: toolboxes/gtplus/algorithm/gtPlusSPIRIT2DTOperator.h [hunk 15]
```diff
@@ -334,8 +337,11 @@ bool gtPlusSPIRIT2DTOperator<T>::obj(const hoNDArray<T>& x, T& obj)
 
             GADGET_CHECK_RETURN_FALSE(Gadgetron::multipleMultiply(currComplexIm, curr_forward_kernel, this->res_after_apply_kernel_));
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRIT2DTOperator.h]: What is "multipleMultiply", sounds like something that should not be here.

File: toolboxes/gtplus/algorithm/gtPlusSPIRITNoNullSpace2DTOperator.h [hunk 16]
```diff
@@ -207,8 +207,11 @@ bool gtPlusSPIRITNoNullSpace2DTOperator<T>::grad(const hoNDArray<T>& x, hoNDArra
 
             GADGET_CHECK_RETURN_FALSE(Gadgetron::multipleMultiply(currComplexIm, curr_adjoint_forward_kernel, this->res_after_apply_kernel_));
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITNoNullSpace2DTOperator.h]: Again, what is this?

File: toolboxes/gtplus/algorithm/gtPlusSPIRITNoNullSpaceOperator.h [hunk 17]
```diff
@@ -83,7 +83,7 @@ bool gtPlusSPIRITNoNullSpaceOperator<T>::grad(const hoNDArray<T>& x, hoNDArray<T
 
         // apply kernel and sum
         GADGET_CHECK_RETURN_FALSE(Gadgetron::multipleMultiply(this->complexIm_, *this->adjoint_forward_kernel_, this->res_after_apply_kernel_));
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::sumOverSecondLastDimension(this->res_after_apply_kernel_, this->res_after_apply_kernel_sum_over_));
+        GADGET_CHECK_RETURN_FALSE(this->sumOverSecondLastDimension(this->res_after_apply_kernel_, this->res_after_apply_kernel_sum_over_));
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITNoNullSpaceOperator.h]: More of this stuff.

File: toolboxes/gtplus/algorithm/gtPlusSPIRITNoNullSpaceOperator.h [hunk 18]
```diff
@@ -113,7 +113,7 @@ bool gtPlusSPIRITNoNullSpaceOperator<T>::obj(const hoNDArray<T>& x, T& obj)
 
         // apply kernel and sum
         GADGET_CHECK_RETURN_FALSE(Gadgetron::multipleMultiply(this->complexIm_, *this->forward_kernel_, this->res_after_apply_kernel_));
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::sumOverSecondLastDimension(this->res_after_apply_kernel_, this->res_after_apply_kernel_sum_over_));
+        GADGET_CHECK_RETURN_FALSE(this->sumOverSecondLastDimension(this->res_after_apply_kernel_, this->res_after_apply_kernel_sum_over_));
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITNoNullSpaceOperator.h]: again

File: toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h [hunk 19]
```diff
@@ -253,7 +253,7 @@ bool gtPlusSPIRITOperator<T>::forwardOperator(const hoNDArray<T>& x, hoNDArray<T
             Gadgetron::multipleMultiply(complexIm_, *forward_kernel_, res_after_apply_kernel_);
         }
 
-        Gadgetron::sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_);
+        this->sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_);
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h]: again

File: toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h [hunk 20]
```diff
@@ -292,7 +292,7 @@ bool gtPlusSPIRITOperator<T>::adjointOperator(const hoNDArray<T>& x, hoNDArray<T
 
             // apply kernel and sum
             Gadgetron::multipleMultiply(complexIm_, *adjoint_kernel_, res_after_apply_kernel_);
-            Gadgetron::sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_);
+            this->sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_);
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h]: again

File: toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h [hunk 21]
```diff
@@ -332,7 +332,7 @@ bool gtPlusSPIRITOperator<T>::computeRighHandSide(const hoNDArray<T>& x, hoNDArr
             GADGET_CHECK_RETURN_FALSE(Gadgetron::multipleMultiply(complexIm_, *forward_kernel_, res_after_apply_kernel_));
         }
 
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_));
+        GADGET_CHECK_RETURN_FALSE(this->sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_));
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h]: again

File: toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h [hunk 22]
```diff
@@ -372,7 +372,7 @@ bool gtPlusSPIRITOperator<T>::grad(const hoNDArray<T>& x, hoNDArray<T>& g)
 
         // apply kernel and sum
         GADGET_CHECK_RETURN_FALSE(Gadgetron::multipleMultiply(complexIm_, *adjoint_forward_kernel_, res_after_apply_kernel_));
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_));
+        GADGET_CHECK_RETURN_FALSE(this->sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_));
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h]: again

File: toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h [hunk 23]
```diff
@@ -409,7 +409,7 @@ bool gtPlusSPIRITOperator<T>::obj(const hoNDArray<T>& x, T& obj)
 
         // apply kernel and sum
         GADGET_CHECK_RETURN_FALSE(Gadgetron::multipleMultiply(complexIm_, *forward_kernel_, res_after_apply_kernel_));
-        GADGET_CHECK_RETURN_FALSE(Gadgetron::sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_));
+        GADGET_CHECK_RETURN_FALSE(this->sumOverSecondLastDimension(res_after_apply_kernel_, res_after_apply_kernel_sum_over_));
```
Comment (hansenms) [inline toolboxes/gtplus/algorithm/gtPlusSPIRITOperator.h]: again

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.cpp [hunk 24]
```diff
@@ -23,123 +23,6 @@ namespace Gadgetron {
...
-        {
-            GERROR_STREAM("Errors in sumOverSecondLastDimension(const hoNDArray<T>& x, hoNDArray<T>& r) ... ");
-            return false;
-        }
-        return true;
-    }
-
     // e.g. x is 3D and y is 4D array, r(:,:,:,n) = y(:,:,:,n) .* x
     template<typename T> 
     bool multiplyOverLastDimension(const hoNDArray<T>& x, const hoNDArray<T>& y, hoNDArray<T>& r)
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.cpp]: This function should not be here, or anywhere.

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.cpp [hunk 25]
```diff
@@ -2283,16 +1474,6 @@ namespace Gadgetron {
...
-    template EXPORTGTPLUS bool sumOverLastDimension(const hoNDArray<double>& x, hoNDArray<double>& r);
-    template EXPORTGTPLUS bool sumOverLastDimension(const hoNDArray< std::complex<float> >& x, hoNDArray< std::complex<float> >& r);
-    template EXPORTGTPLUS bool sumOverLastDimension(const hoNDArray< std::complex<double> >& x, hoNDArray< std::complex<double> >& r);
-
-    template EXPORTGTPLUS bool sumOverSecondLastDimension(const hoNDArray<float>& x, hoNDArray<float>& r);
-    template EXPORTGTPLUS bool sumOverSecondLastDimension(const hoNDArray<double>& x, hoNDArray<double>& r);
-    template EXPORTGTPLUS bool sumOverSecondLastDimension(const hoNDArray< std::complex<float> >& x, hoNDArray< std::complex<float> >& r);
-    template EXPORTGTPLUS bool sumOverSecondLastDimension(const hoNDArray< std::complex<double> >& x, hoNDArray< std::complex<double> >& r);
-
     template EXPORTGTPLUS bool multiplyOverLastDimension(const hoNDArray<float>& x, const hoNDArray<float>& y, hoNDArray<float>& r);
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.cpp]: All these functions should not exist

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.cpp [hunk 26]
```diff
@@ -2303,31 +1484,6 @@ namespace Gadgetron {
...
-    template EXPORTGTPLUS bool sumOver4thDimension(const hoNDArray<double>& x, hoNDArray<double>& r);
-    template EXPORTGTPLUS bool sumOver4thDimension(const hoNDArray< std::complex<float> >& x, hoNDArray< std::complex<float> >& r);
-    template EXPORTGTPLUS bool sumOver4thDimension(const hoNDArray< std::complex<double> >& x, hoNDArray< std::complex<double> >& r);
-
-    template EXPORTGTPLUS bool sumOver5thDimension(const hoNDArray<float>& x, hoNDArray<float>& r);
-    template EXPORTGTPLUS bool sumOver5thDimension(const hoNDArray<double>& x, hoNDArray<double>& r);
-    template EXPORTGTPLUS bool sumOver5thDimension(const hoNDArray< std::complex<float> >& x, hoNDArray< std::complex<float> >& r);
-    template EXPORTGTPLUS bool sumOver5thDimension(const hoNDArray< std::complex<double> >& x, hoNDArray< std::complex<double> >& r);
-
     template EXPORTGTPLUS bool multiplyOver3rdDimension(const hoNDArray<float>& x3D, const hoNDArray<float>& y4D, hoNDArray<float>& r);
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.cpp]: Same with these, should not exist

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.h [hunk 27]
```diff
@@ -236,36 +224,6 @@ namespace Gadgetron {
     template<typename T> EXPORTGTPLUS bool divideOverLastDimension(const hoNDArray<T>& x, const hoNDArray<T>& y, hoNDArray<T>& r);
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.h]: Should not exists

File: toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.h [hunk 28]
```diff
@@ -334,46 +292,6 @@ namespace Gadgetron {
     template<typename T> EXPORTGTPLUS bool stdOver3rdDimension(const hoNDArray<T>& x, hoNDArray<T>& std, bool NMinusOne);
```
Comment (hansenms) [inline toolboxes/gtplus/workflow/gtPlusISMRMRDReconUtil.h]: Should not exist

**Comment by hansenms** [comment]
There are still multiple of these functions specific to a given dimension. And the couple of places that you have taken them out, you have simply inlined them. Not quite sure I understand what you are trying to do here??

**Comment by gtdeveloper5** [comment]
More changes are made to remove all remaining sumOver, multipleOver and other specific dimension calls. The purpose here is to set up a stage that i can start to put down the 2d grappa functions in the mri_core toolbox.

**Comment by hansenms** [comment]
Why are you inlining your permutes instead of using the permute function?

**Comment by gtdeveloper5** [comment]
All previously class member permute functions are removed and replaced with calls to general permute functions, with the commit "remove remaining permuteRO calls".

**Comment by hansenms** [comment]
Line 938 of gtPlusGRAPPA.h you have simply inlined the permute as I said. There is a comment on it. It is also repeated in gtPlusSPIRIT.h. Why are you not using the permute functions?

You also still have a "performSumOverSrcChannel" function.

**Comment by gtdeveloper5** [comment]
The general permute is used to replace lines in grappa and spirit for 3d kernel estimation. 

The function performSumOverSrcChannel is using general sum_over_dimension for its implementation. This function is a class member and added to avoid duplicates when calling  sum_over_dimension in multiple places.

**Comment by inati** [comment]
I don't understand that comment at all.  What do you mean avoid duplicates?

**Comment by gtdeveloper5** [comment]
Because there are more than one line needed to call sum_over_dimension, this performSumOverSrcChannel member function is added. So those lines will not be written several times in multiple places.

**Comment by hansenms** [comment]
Looks better now. Probably as much mess as we can clean up with this one. @inati, are you OK with merging?

**Comment by inati** [comment]
Good for now.

**Pull request merged**